kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -1050,42 +1048,35 @@ func (c *IPAMController) syncIPAM() error {
 	// Delete any nodes that we determined can be removed in checkAllocations. These
 	// nodes are no longer in the Kubernetes API, and have no valid allocations, so can be cleaned up entirely
 	// from Calico IPAM.
-	var storedErr error
-	if len(nodesToRelease) > 0 {
-		log.WithField("num", len(nodesToRelease)).Info("Found a batch of nodes to release")
-	}
-	for _, cnode := range nodesToRelease {
-		logc := log.WithField("node", cnode)
-
-		// Potentially rate limit node cleanup.
-		rlKey := rateLimiterItemKey{Type: RateLimitCalicoDelete, Name: cnode}
-		time.Sleep(c.rl.When(rlKey))
-		logc.Info("Cleaning up IPAM affinities for deleted node")
-		if err := c.cleanupNode(cnode); err != nil {
-			// Store the error, but continue. Storing the error ensures we'll retry.
-			logc.WithError(err).Warnf("Error cleaning up node")
-			storedErr = err
-			continue
-		}
-		c.rl.Forget(rlKey)
-	}
-	if storedErr != nil {
-		return storedErr
+	if err = c.releaseNodes(nodesToRelease); err != nil {
+		return err
 	}
 
 	c.allocationState.syncComplete()
 	log.Debug("IPAM sync completed")
+
+	// If there is still dirty state, then we need to do another pass.
+	if len(c.confirmedLeaks) > 0 {
+		log.WithField("num", len(c.confirmedLeaks)).Info("Confirmed leaks still exist, scheduling another pass")
+		kick(c.syncChan)
+	}
 	return nil
 }
 
 // garbageCollectKnownLeaks checks all known allocations and garbage collects any confirmed leaks.
 func (c *IPAMController) garbageCollectKnownLeaks() error {
+	defer logIfSlow(time.Now(), "Leak GC complete")
+
+	// limit the number of concurrent IPs we attempt to release at once.
+	maxBatchSize := 10000
+
+	var opts []ipam.ReleaseOptions
+	leaks := map[ipam.ReleaseOptions]*allocation{}
 	for id, a := range c.confirmedLeaks {
 		logc := log.WithFields(a.fields())
 
-		// Final check that the allocation is leaked, this time ignoring our cache
-		// to make sure we're working with up-to-date information.
-		if c.allocationIsValid(a, false) {
+		// Final check that the allocation is leaked.
+		if c.allocationIsValid(a, true) {
```

Comment:
This is one change I'd like a second opinion on - it was taking a LOT of time to query each Pod individually, and switching to a cache check will potentially reduces hundreds to thousands of API calls made by this controller when - for example - tearing down a large number of nodes at once.

Reply:
But, if for some reason our informer cache doesn't match the actual state of the cluster, then we may take actions that are incorrect (like releasing an IP that we don't need to). I think this is unlikely though.

Reply:
I'm not very familiar with this, so I went digging into the podLister that we use as cache (please let me know if I'm way off), but it seems like it does a k8s watch on pods? So I do agree with you, seems pretty unlikely that we'd do something wrong here due to the cache being outdated...